### DISCUSSION TITLE
Run each endpoint as background tasks in individual working directories at same hierarchy

### DISCUSSION THREAD

**Discussion by yarnabrina** [fastapi/fastapi#9352] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
"""Define FastAPI server for Minimal Forecast SDK."""
import logging
import os
import pathlib
import shutil
import time
import uuid

import fastapi
import fastapi.encoders
import fastapi.responses
import pydantic
import uvicorn

logging.basicConfig(
    format="{asctime} {levelname} {name} {funcName} {lineno} {message}",
    datefmt="%Y-%m-%dT%H:%M:%S",
    style="{",
    level=logging.INFO,
)

JsonResponse = fastapi.responses.JSONResponse


def get_json_response(response: dict) -> JsonResponse:
    return fastapi.responses.JSONResponse(fastapi.encoders.jsonable_encoder(response))


def setup_working_directory(identifier: str) -> None:
    updated_working_directory = pathlib.Path(identifier)
    updated_working_directory.mkdir()

    os.chdir(updated_working_directory)


def teardown_working_directory(identifier: str) -> None:
    working_directory_parent = pathlib.Path().resolve().parent
    os.chdir(working_directory_parent)

    shutil.rmtree(identifier)


def time_taking_function(identifier: str, duration: int) -> None:
    logging.info(f"Before setup for identifier={identifier}")
    setup_working_directory(identifier)
    logging.info(f"After setup for identifier={identifier}")

    time.sleep(duration)

    logging.info(f"Before teardown for identifier={identifier}")
    teardown_working_directory(identifier)
    logging.info(f"After teardown for identifier={identifier}")


class MainEndpointParameters(pydantic.BaseModel):
    duration: int
    identifier: str | None = None


APP = fastapi.FastAPI()


@APP.get("/health")
def health_endpoint() -> JsonResponse:
    return get_json_response({"status": "active"})


@APP.post("/main")
def main_endpoint(
    parameters: MainEndpointParameters, background_tasks: fastapi.BackgroundTasks
) -> JsonResponse:
    logging.info(f"Received duration={parameters.duration}")

    if not (identifier := parameters.identifier):
        identifier = str(uuid.uuid4())
    logging.info(f"Execution identifier={identifier}")

    background_tasks.add_task(time_taking_function, identifier, parameters.duration)

    return get_json_response({"status": "pending", "identifier": identifier})


if __name__ == "__main__":
    uvicorn.run("server:APP", host="0.0.0.0", port=8000, log_level="debug", reload=True)
```

### Description

* Call the "/main" endpoint multiple times in quick succession, where "quick" means before the execution of previous executions get over.
* Directories for each request are created in a nested fashion.
* If one of the earlier request finishes first and performs the "teardown", then next requests fail to do necessary cleanup (and in actual problem, fails to do anything as looses relevant data)

For example, copy-paste all of these together in the terminal:

```bash
curl --request POST localhost:8000/main --header 'Content-Type: application/json' --data '{"duration": 15, "identifier": "a"}'
curl --request POST localhost:8000/main --header 'Content-Type: application/json' --data '{"duration": 15, "identifier": "b"}'
curl --request POST localhost:8000/main --header 'Content-Type: application/json' --data '{"duration": 15, "identifier": "c"}'
```

After receiving the 3 "pending" responses, this will be folder contents:

```console
❯ ls -Rl
total 8
drwxr-xr-x  3 aniray1  staff    96 Apr  5 22:29 __pycache__
drwxr-xr-x  3 aniray1  staff    96 Apr  5 23:00 a
-rw-r--r--  1 aniray1  staff  2274 Apr  5 22:29 server.py

./__pycache__:
total 8
-rw-r--r--  1 aniray1  staff  2790 Apr  5 22:29 server.cpython-310.pyc

./a:
total 0
drwxr-xr-x  3 aniray1  staff  96 Apr  5 23:00 b

./a/b:
total 0
drwxr-xr-x  2 aniray1  staff  64 Apr  5 23:00 c

./a/b/c:
total 0
```

And after "a" gets completed, "b" and "c" will fail, with similar error as below:

```console
FileNotFoundError: [Errno 2] No such file or directory: 'c'
```

What I expected/wanted to happen was to have "a", "b", "c" folders to be created at "root" level, i.e. at same level as "server.py". Then these 3 tasks would be totally independent of each other.

My question is how to achieve the above expectation.

### Operating System

macOS

### FastAPI Version

0.95.0

### Python Version

3.10.10

**Comment by jgould22** [comment, marked as answer]
You are calling `os.chdir(path)` which is changing the path and subsequent requests are now rooted in that folder.

Dont do that.

Create a data directory on startup to do your work in.

Then use the full file path to operate on that directory.

**Reply by yarnabrina** [reply]
I do understand that `os.chdir` is causing it. I know for sure there will be concurrent requests, and each of them will take time to finish. I don't want those to affect each other, and I want all different calls to be completely separate. I don't think separation can be ensured with a common root directory. If I am missing something, can you please elaborate?

**Reply by jgould22** [reply]
Put them in a different directory so your requests don't collide with the pycache and server.py (and whatever else is going on in that directory) and for organization(so if your server stops during a request and doesnt manage to clean up you only have one place to clear).

Then do something like this (untested, but you get the idea)
```python

def get_request_identifier(parameters: MainEndpointParameters) -> str:

     if not (identifier := parameters.identifier):
        identifier = str(uuid.uuid4())

     return identifier


def setup_working_directory(identifier = Depends(get_request_identifier)) -> str:

    updated_working_directory = pathlib.Path(identifier)
    updated_working_directory.mkdir()

    return updated_working_directory


def teardown_working_directory(working_dir: str) -> None:

    shutil.rmtree(working_dir)


def time_taking_function(identifier: str, duration: int, working_dir: str) -> None:

    time.sleep(duration)
    Path(f'{working_dir}/some_file.txt').touch()

    logging.info(f"Before teardown for identifier={identifier}")
    teardown_working_directory(working_dir)
    logging.info(f"After teardown for identifier={identifier}")


class MainEndpointParameters(pydantic.BaseModel):
    duration: int
    identifier: str | None = None

APP = fastapi.FastAPI()

@APP.get("/health")
def health_endpoint() -> JsonResponse:
    return get_json_response({"status": "active"})


@APP.post("/main")
def main_endpoint(
    identifier = Depends(get_request_identifier),
    working_dir: str = Depends(setup_working_directory),
    background_tasks: fastapi.BackgroundTasks
) -> JsonResponse:

    background_tasks.add_task(time_taking_function, identifier, parameters.duration, working_dir)

    return get_json_response({"status": "pending", "identifier": identifier})


if __name__ == "__main__":
    uvicorn.run(APP, host="0.0.0.0", port=8000, log_level="debug", reload=True)
```

**Reply by yarnabrina** [reply]
Thanks for quick response. A quick clarification: how will this ensure `time_taking_function` runs in desired working directory of `{root}/{identifier}`?

`time.sleep(duration)` part is still running in `{root}`, right? If so, that's not possible for my actual use case, mainly because of the package I'm using store some files in PWD as checkpoints and doesn't provide ways to configure locations.
